relays/substrate/src/headers_maintain.rs

```diff
@@ -153,6 +159,11 @@ where
 		})
 		.await;
 
+		// if justifications subscription has been dropped, resubscribe
+		if justifications.stream.is_none() {
```

Comment:
Is there a place in code that should be setting `justifications.stream = None`? I couldn't find anything, which means that the stream can be `None` only initially, no?

Reply:
only returns None when the stream has been terminated.

Reply:
> Is there a place in code that should be setting justifications.stream = None

It may have been set to `None` if first subscription has failed. But it was not enough ofc - I've now added reconnect when subscription has been dropped after initiation.

Reply:
@svyatonik sorry I forgot to mention that an annoyance is https://github.com/paritytech/jsonrpsee/issues/220

Basically if you unsubscribe to a subscription the client won't register the request and if the server responds with the request ID it will most likely fail, as `invalid requestID` (which I think `jsonrpc server pubsub` does). Fail means closing the connection and it has to restarted :(

The reason for this is that `jsonrpsee ws server` didn't include request ID in responses for subscription from the beginning we will fix it soon.

Reply:
@niklasad1 Thanks! :) So if I understand correctly, until that fix we'll be disconnecting from the server if there are too many unread notifications? That's ok - iiuc this disconnect will be reported as `Error::Internal(_)` when we'll try to use connection next time && we're now able to process it properly.

Reply:
yeah almost,

1. The subscription capacity gets full because of too many unread notifications
2. If you then drop the entire subscription as I think you do
3. client send `unsubscribe_to_some_method, id: x`
4. server ACKs `unsubscribe_to_some_method` the request with `id: x` 
5. client doesn't know what `id: x`  is and closes the connection because of `Error::InvalidRequestId`.

A super hacky-workaround is to keep the all subscriptions but this means that server will send out needless messages that are not processed or make the buffer huge for now, I don't think you will run into this but you might ^^

I can ping you when the fix is merged but I haven't tested against `jsonrpc ws server` FWIW.

Reply:
Yes, please ping me :) I'll be able to test if it works with substrate (jsonrpc) as well

Reply:
@svyatonik 

Published 0.2.0-alpha.2, it has some breaking changes renaming some fields.
Also I think it will fix the compilation warnings generated the proc macros.

Reply:
Thank you - will try to check it soon :)

Reply:
@niklasad1 I've opened #828 - could confirm that there's no warnings + connection is not lost when we unsubscribe. Thank you! :)